BootloaderCorePkg/Library/S3SaveRestoreLib/S3SaveRestore.c

```diff
@@ -0,0 +1,160 @@
+/** @file
+
+  Copyright (c) 2017, Intel Corporation. All rights reserved.<BR>
+  SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+
+#include <PiPei.h>
+#include <Library/DebugLib.h>
+#include <Library/HobLib.h>
+#include <Library/BaseMemoryLib.h>
+#include <Library/S3SaveRestoreLib.h>
+#include <Library/MpInitLib.h>
+#include <Library/IoLib.h>
+#include <Guid/SmmInformationGuid.h>
+#include <Library/BootloaderCommonLib.h>
+
+
+/**
+  This function uploads the register information in TSEG area
+  designated for S3 save/restore purpose.
+  SMMBASE_INFO needs to be present at the starting of TSEG region.
+  Other regions can follow in any order.
+  The 'Val' field of the REG_INFO register information is
+  updated by the payload.
+
+  @param    SmmBaseInfo   Pointer to TSEG region
+  @param    DataPtr       Address of the structure to be copied to TSEG
+
+  @retval   None
+**/
+VOID
+EFIAPI
+S3SetData (
+  IN  VOID    *SmmBaseInfo,
+  IN  UINT32  SmmSize,
+  IN  VOID    *DataPtr
+  )
+{
+  UINT8             *Ptr;
+
+  if ( ((BL_PLD_COMM_HDR *)SmmBaseInfo)->Signature != SMM_BASE_INFO_SIG) {
+    DEBUG((DEBUG_ERROR, "Add SMMBASE_INFO region first\n"));
+    return;
+  }
```

Comment:
we don't need check there. Let the platform to set SmmBase info firstly if the platform wants to make it as the first record.